Alternative For Selecting Image Comparison

danieldresser-ie opened the pull request:
This is a hacky experiment, but I think the behaviour is something like it should be?

Aside from the various WIP code for this feature, I should also mention that I had to hack the recent change to OSLCode in order to get it compiling at IE - I swapped out the smart pointer for a raw pointer in the dictionary key, since the smart pointer wasn't supported ( maybe we're on older boost ).  Is it actually desirable that this dictionary could force one of the keys to stay alive?

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1260,10 +1260,11 @@ def __setValue( self, value, *unused ) :
 
 		if turningOn and len( viewNames ) > 1:
+			node = self.getPlug().node()
 			if node["compare"]["view"].getValue() == node["view"].getValue() or not node["compare"]["view"].getValue() in viewNames :
 				# We're turning on comparison, but currently the comparison view is identical to the
 				# default view - pick a different view, so we actually get a meaningful comparison
```
Based on this comment, it seems like all this logic is obsolete now anyway. We're not using different views for comparison, and there's not even a UI for the user to choose a view for comparison. So I think we want the comparison view to always be the same as the main view don't we?

johnhaddon commented on python/GafferUI/CompoundEditor.py:
```diff
@@ -1682,7 +1680,7 @@ def __getBookmarkSet( self ) :
 
 	def __leave( self, widget ) :
-
+		# TODO - does this function do anything?
```
It doesn't look like it. I think it should have been removed completely in 0658b4c1042b0fa9db28739539142d3e1aefb07b, which removed the one bit of it that used to do anything. Let's get rid.

johnhaddon commented on include/GafferOSL/OSLCode.h:
```diff
@@ -88,7 +88,7 @@ class GAFFEROSL_API OSLCode : public OSLShader
 
 		ShaderCompiledSignal m_shaderCompiledSignal;
-		std::unordered_map<Gaffer::ConstGraphComponentPtr, Gaffer::Signals::ScopedConnection> m_nameChangedConnections;
+		std::unordered_map<const Gaffer::GraphComponent*, Gaffer::Signals::ScopedConnection> m_nameChangedConnections;
```
What compilation error do you get at IE? Ideally IE would be building with the same libs/options that we do, so there aren't any surprises. 

That said, I _think_ I was probably just being lazy in using a Ptr here. I don't think it's possible for the plug to die while it's still in this map, because to die it needs to be removed from its parent, and when that happens we remove it from the map in `parameterRemoved()`. Does that make sense?

johnhaddon commented on python/GafferUI/_StyleSheet.py:
```diff
@@ -1638,6 +1638,20 @@ def styleColor( key ) :
 		background: $background;
 	}
 
+	QFrame[gafferClass="GafferImageUI.ImageViewUI._CompareImageWidget"] #menuDownArrow {
```
I think the reason the original styling isn't getting inherited is that it is based on an exact match for the most derived class in Gaffer. Instead you can do a match against the base classes too, as documented here : https://github.com/GafferHQ/gaffer/blob/main/python/GafferUI/_StyleSheet.py#L203.

johnhaddon reviewed (commented):
> This is a hacky experiment, but I think the behaviour is something like it should be?

Yeah, the end-user behaviour feels pretty reasonable. The menu does feel quite complex, and it definitely needs a "Comparison Image" title, but it does seem a natural way of providing some useful functionality.

I think it's worth getting the simple fixes merged right away, so we can focus down onto the hacks. About those...

- Adding the `ImageView.compare.image` plug does seem consistent with the way things are at the moment, where `ImageView.in` is also an ImagePlug, and the translation from NodeSet to plug is performed externally (by the Viewer in that case). But is it strange to split this translation responsibility so that in one case it lies with the Viewer and in the other it lies with ImageViewUI?
- Longer term, and with https://github.com/GafferHQ/gaffer/issues/50 in mind, I'm less sure about how the NodeSet should be communicated to the View. #50 requires that the whole UI be serialisable, which means all the state being provided by plugs. So does that mean that there should be a `StringPlug nodeSetPlug()` on the View? Or would that be on the Viewer still, and it would do the translation to a connection still? Or some other mechanism (all suggestions welcome)?
- Having the catalogue outputs be multiple individual nodes inside the view that are then referenced by individual sets outside of the view feels like the worst bit of this to me. Could we just keep `compare.catalogueOutput` as it is? In theory I might want my comparison node set to specify a catalogue and then use `catalogueOutput` on top of that, so I don't think they need to be mutually exclusive. We might want to present them as mutually exclusive for simplicity in the UI, but for simplicity of implementation we could just have both?
- _CompareDummyNodeSetEditor is also a bit hacky. Is there a better abstraction for something a _PinningWidget could operate on? Could that be the `StringPlug nodeSetPlug()` I was musing about above? Or something else?

Depending on how valued this feature is to users, I could probably tolerate a degree of hackery for now to get it implemented. But it'd also be good to spend a bit of time trying to figure out how it might relate to #50, and if there mightn't be a cleaner way of expressing it...

danieldresser-ie commented:
Completely rewritten - probably not finished yet, but definitely ready for another look.

johnhaddon commented:
> probably not finished yet, but definitely ready for another look

If it's not finished yet, and there isn't something specific you want me to look at, I'll just wait until it's done.

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1158,6 +1166,11 @@ def __init__( self, plug, **kw ) :
 
 		self._updateFromPlug()
 
+		self.dragEnterSignal().connectFront( Gaffer.WeakMethod( self.__dragEnter ), scoped = False )
+		self.dragLeaveSignal().connectFront( Gaffer.WeakMethod( self.__dragLeave ), scoped = False )
+		self.dropSignal().connectFront( Gaffer.WeakMethod( self.__drop ), scoped = False )
```
I presume `connectFront()` is needed so that we get in before PlugValueWidget's own handlers? If so, could we comment that - it took me a little while to figure it out.

danieldresser-ie replied:
I had actually copied the connectFront() carelessly from somewhere else that uses dragEnterSignal when I was just trying to get it working.  It's not absolutely necessary to get this working ( since the drags from nodes in the graph don't originate from PlugValueWidgets, PlugValueWidgets's handler doesn't grab them ).  But I think in light of your comment about returning True from  __dragEnter, I think it sounds like we should do a connectFront here, and I'll add a comment to reflect that?

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1167,6 +1180,43 @@ def _updateFromPlug( self ) :
+			if event.data.size() and isinstance( event.data[0], Gaffer.Node ) :
+				result = True
+
+		if result :
+			for i in self.__row:
+				i.setHighlighted( True )
```
I do like the bigger drop target provided by `_CompareParentPlugValueWidget`, but I think it might be clearer if we just highlighted the "pinning" widget. The same way a whole StandardNodeGadget can be the drop target for a connection, but we highlight the specific input plug that will be affected.

danieldresser-ie replied:
Addressed.

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1167,6 +1180,43 @@ def _updateFromPlug( self ) :
+			for i in self.__row:
+				i.setHighlighted( True )
+
+		return result
```
I think we probably want to `return True` here, so we never have to worry about weird interactions with the drag handling from PlugValueWidget.

danieldresser-ie replied:
I'm not aware of other places in Gaffer where we make a drag event return True in the case where we don't actually want to handle the drop, but we want to prevent the default handler from accepting it.

I've done this by returning True on invalid drags, but not setting highlighted.  Does this seem right?

danieldresser-ie replied:
Another note here:  I'm now treating drags of nodes with no valid image plugs the same as drags of things that aren't nodes:  I return True to suppress further processing, but don't highlight or accept the drop.  Does that seem right?

johnhaddon replied:
That seems reasonable to me

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1167,6 +1180,43 @@ def _updateFromPlug( self ) :
+	def __dragEnter( self, tabbedContainer, event ) :
+
+		result = False
+		if isinstance( event.data, Gaffer.Node ) :
+			result = True
+		elif isinstance( event.data, Gaffer.Set ) :
+			if event.data.size() and isinstance( event.data[0], Gaffer.Node ) :
+				result = True
```
Rather than repeat this logic in `__drop()`, perhaps we could stick it in a `__dropNode()` function that takes the event and returns the node or `None`?

danieldresser-ie replied:
Addressed.

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1300,49 +1350,217 @@ def __toggle( self, *args ):
+			self.__icon._qtWidget().setFixedWidth( 13 )
+			self.__icon.buttonPressSignal().connect( Gaffer.WeakMethod( self.__showEditorFocusMenu ), scoped=False )
+
+			# TODO - why does the pinning dropdown use collapsibleArrowDown instead of menuIndicator?
```
I don't know - perhaps give it a try, and if it seems to work, change it?

danieldresser-ie replied:
Added commit feef55b

Looks good to me, and Murray thought it was a good idea.

johnhaddon replied:
We still need to remove the TODO from ImageViewUI.

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1300,49 +1350,217 @@ def __toggle( self, *args ):
+		self.update()
 
-		GafferUI.PlugValueWidget.__init__( self, self.__button, plug, **kw )
+	# TODO - hotkeys?
```
I think we're probably fine to omit hotkeys for this.

danieldresser-ie replied:
Removed comment.

johnhaddon commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1300,49 +1350,217 @@ def __toggle( self, *args ):
+		if self.__nodeSet is self.__defaultNodeSet:
+			compareImage = self.__node["in"]
+		elif len( self.__nodeSet ):
+			for plug in Gaffer.Plug.RecursiveOutputRange( self.__nodeSet[-1] ) :
```
Could use `GafferImage.ImagePlug.RecursiveOutputRange` here.

danieldresser-ie replied:
Addressed.

johnhaddon reviewed (commented):
Hey Daniel, I've put up a hasty review with the time I had left, since you indicated offline that this was ready enough for a look. In use I like it - feels good. I've made some comments inline that are hopefully helpful, or at least make sense...

danieldresser-ie commented on python/GafferUI/Widget.py:
```diff
@@ -316,12 +316,20 @@ def setHighlighted( self, highlighted ) :
 			return
 
-		self._qtWidget().setProperty( "gafferHighlighted", GafferUI._Variant.toVariant( highlighted ) )
+		widget = self._qtWidget()
+		if type( widget ) == QtWidgets.QWidget :
+			widget = widget.layout().itemAt( 0 ).widget()
```
OK, if the convention is that all widgets that call the base constructor with anything not derived from QWidget are responsible for overriding setHighlighted, then the issue is that GafferUI.Button fails to do this, and we don't need to change Widget.

This now irrelevant to this PR, since you're happy to only highlight _CompareImageWidget, which derives from Frame, which passes a QWidget derived QFrame to the constructor, so is unaffected by this.

Let me know if you want me to put together a separate PR with a fix for this, with the stylesheet change to allow buttons to be highlighted ... not sure whether there is another use for highlighting buttons ... I guess our API sort of implies it should work.

johnhaddon replied:
> OK, if the convention is that all widgets that call the base constructor with anything not derived from QWidget are responsible for overriding setHighlighted, then the issue is that GafferUI.Button fails to do this, and we don't need to change Widget.

I don't understand this - `GafferUI.Button` passes `QPushButton` to `Widget.__init__`, so wouldn't hit the new code in Widget anyway. Button doesn't need to override `setHighlighted` - we just need a stylesheet for `QPushButton` with the `gafferHighlighted` property. It would be `_CompareModePlugValueWidget` that would have needed to override `setHighlighted`.

> Let me know if you want me to put together a separate PR with a fix for this, with the stylesheet change to allow buttons to be highlighted

I don't think it's urgent - let's just do it when we actually need it.

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1300,49 +1350,217 @@ def __toggle( self, *args ):
+		return False
+	"""
+
+	def update( self, *unused ) :
```
Made _update protected.

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1300,49 +1350,217 @@ def __toggle( self, *args ):
+		try :
+			self.__icon.setImage( icon )
+		except Exception as e :
+			sys.stderr.write( str(e) )
```
It feels like it would be hard to guarantee that no users of Image are expecting to get an exception when it fails to construct ( and what state is that supposed to leave it in? ).  A reasonable compromise seems to be moving the code from CompoundEditor ( where I copied it from ) to Button ( as well as switching to msg() ).

johnhaddon replied:
> It feels like it would be hard to guarantee that no users of Image are expecting to get an exception when it fails to construct

It feels like a pretty reasonable assumption to me. What use cases are you envisaging where the exception handling would be any more than printing a message and carrying on?

> and what state is that supposed to leave it in?

Showing the warning icon maybe? Or just nothing? The reason I was thinking in this direction was that we recently had a case where Gaffer was completely crippled because a custom tool was missing its icon, and an unhandled exception prevented the UI from loading...

> A reasonable compromise seems to be moving the code from CompoundEditor

That's better, and I'll merge it, but I still think there's a good case for doing this even more centrally.

danieldresser-ie commented on src/GafferImageUI/ImageView.cpp:
```diff
@@ -1601,10 +1620,14 @@ void ImageView::insertConverter( Gaffer::NodePtr converter )
 	NodePtr preprocessor = getPreprocessor();
 	Plug::OutputContainer outputsToRestore = preprocessor->getChild<ImagePlug>( "in" )->outputs();
 
+	// TODO - this is now actually broken, because any converters should be downstream of the
+	// comparison switch, so they affect both sides of the wipe.  Hopefully John has an idea on how
+	// to approach this.
```
Oh, OK.  Maybe it's fine for how this actually gets used.  I was picturing a converter with an image input, where it would be weird to have it only applied to one side of the wipe, but maybe there's no use case for this ... we certainly have better ways to apply color transforms.

danieldresser-ie commented on python/GafferUI/Widget.py:
```diff
@@ -316,12 +316,20 @@ def setHighlighted( self, highlighted ) :
+		if "gafferHighlighted" not in widget.dynamicPropertyNames() :
 			return False
 
 		return GafferUI._Variant.fromVariant( self._qtWidget().property( "gafferHighlighted" ) )
```
I've now omitted all this from this PR, since it no longer requires highlighting buttons.

danieldresser-ie commented on python/GafferImageUI/ImageViewUI.py:
```diff
@@ -1300,49 +1350,217 @@ def __toggle( self, *args ):
-			)
 
-		return m
+		m.append( "/Catalogue Divider", { "divider" : True, "label" : "Catalogue Outputs" } )
```
Addressed.

danieldresser-ie commented on src/GafferImageUI/ImageView.cpp:
```diff
@@ -1494,9 +1495,26 @@ ImageView::ImageView( const std::string &name )
+	comparisonSwitch->setup( preprocessorInput.get() );
+	comparisonSwitch->indexPlug()->setInput( comparisonContextQuery->valuePlugFromQueryPlug( queryPlug ) );
+	comparisonSwitch->inPlugs()->getChild<ValuePlug>(0)->setInput( preprocessorInput );
+	comparisonSwitch->inPlugs()->getChild<ValuePlug>(1)->setInput( comparisonDeleteContext->outPlug() );
```
Addressed.

danieldresser-ie commented:
I've added a commit that addresses all feedback on ImageView, plus two other commits for the suggested changes to CompoundEditor and Button.

Hopefully this is now getting pretty close, though it still needs a squash if the changes look good, and it would be good to have you take a look at the stylesheet stuff.

johnhaddon commented:
Thanks for the updates Daniel. Couple of minor comments about the comparison widget from playing with it a bit more today :

- I think the comparison menu title would be better as "Comparison Image" - a nouny thing equivalent to the "Editor Focus" title in the main pinning menu.
- It seems inconsistent that the labels for "Follow Numeric Bookmark" are just "1", "2" etc, and the catalogue outputs are "Output 1", "Output 2" etc. Would it be better to just be "1", "2" etc for the outputs too?
- When switching in and output of numeric bookmark mode, the size of the widget jumps, and the change of layout is a bit jarring. Might look a bit rubbish to have a gap for it to grow into though, so maybe this is the best compromise already?

And, as usual, Changes.md needs updating to mention the new feature.

danieldresser-ie commented:
Removed the leftover TODO, changed name to "Comparison Image", and menu item names to "1", "2", ...

Squashed changes, and updated Changes.md

Should be good to go.

johnhaddon commented:
> # TODO - why does the pinning dropdown use collapsibleArrowDown instead of menuIndicator?

I think my memory has dredged up the reason for this. I'm pretty sure @themissingcow did it deliberately because he felt that the bright white menu indicator detracted from the readability of the UI in this case. The important information is contained in the icon on the left, and giving the triangle on the right equal (or greater event) prominence detracted from that. I think he was probably right, and we bring back the fainter icon. I guess either we make a new `menuIndicatorSubtle` icon, or we continue to use `collapsibleDownArrow` cheekily with a comment explaining why. I think it's probably worth doing this before we make the next release...

danieldresser-ie commented:
I don't personally feel strongly, I don't mind if you want to roll back, but for what it's worth, before making this change, I did check in with both Murray and Bea at IE ( who has no knowledge of the history, but does UI stuff ), and both felt that consistency was more valuable than whatever other aesthetic judgements were involved.

If you want to roll it back, do you want me to do the PR for that?  I don't think we need a new name for the icon, but a comment would definitely be good.

johnhaddon commented:
> I did check in with both Murray and Bea at IE ( who has no knowledge of the history, but does UI stuff ), and both felt that consistency was more valuable than whatever other aesthetic judgements were involved.

OK, let's leave it then.
